File: src/main/kotlin/com/huawei/cangjie/descriptors/DiagnosticSink.java
package com.huawei.cangjie.descriptors;

import com.huawei.cangjie.descriptors.rendering.DefaultErrorMessages;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.PsiFile;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public interface DiagnosticSink {
    DiagnosticSink DO_NOTHING = new DiagnosticSink() {
        @Override
        public void report(@NotNull Diagnostic diagnostic) {
        }

        @Override
        public boolean wantsDiagnostics() {
            return false;
        }
    };

    DiagnosticSink THROW_EXCEPTION = new DiagnosticSink() {
        @Override
        public void report(@NotNull Diagnostic diagnostic) {
            if (diagnostic.getSeverity() == Severity.ERROR) {
                PsiFile psiFile = diagnostic.getPsiFile();
                List<TextRange> textRanges = diagnostic.getTextRanges();
                String diagnosticText = DefaultErrorMessages.render(diagnostic);
                throw new IllegalStateException(diagnostic.getFactory().getName() + ": " + diagnosticText + " " + PsiDiagnosticUtils
                        .atLocation(psiFile, textRanges.get(0)));
            }
        }

        @Override
        public boolean wantsDiagnostics() {
            return true;
        }
    };
    interface DiagnosticsCallback {
        void callback(Diagnostic diagnostic);
    }

    void report(@NotNull Diagnostic diagnostic);

    /**
     * use {@link #setCallbackIfNotSet(DiagnosticsCallback)} instead
     * @param callback
     */
    @Deprecated
    default void setCallback(@NotNull DiagnosticsCallback callback) {
        setCallbackIfNotSet(callback);
    }

    default boolean setCallbackIfNotSet(@NotNull DiagnosticsCallback callback) {
        return false;
    }

    default void resetCallback() { }

    boolean wantsDiagnostics();
}

File: src/main/kotlin/com/linqingying/cangjie/diagnostics/DiagnosticSink.kt
package com.linqingying.cangjie.diagnostics

/*

interface DiagnosticSink {
    fun report(diagnostic: Diagnostic)


    interface DiagnosticsCallback {
        fun callback(diagnostic: Diagnostic?)
    }

    companion object {
        val DO_NOTHING: DiagnosticSink = object : DiagnosticSink {
            override fun report(diagnostic: Diagnostic) {
            }

            override fun wantsDiagnostics(): Boolean {
                return false
            }
        }

        val THROW_EXCEPTION: DiagnosticSink = object : DiagnosticSink {
            override fun report(diagnostic: Diagnostic) {
                if (diagnostic.severity == Severity.ERROR) {
                    val psiFile: PsiFile = diagnostic.psiFile
                    val textRanges: List<TextRange> = diagnostic.textRanges
                    val diagnosticText: String =
                        DefaultErrorMessages.render(diagnostic)
                    throw java.lang.IllegalStateException(
                        diagnostic.factory.name + ": " + diagnosticText + " " + PsiDiagnosticUtils
                            .atLocation(psiFile, textRanges[0])
                    )
                }
            }

            override fun wantsDiagnostics(): Boolean {
                return true
            }
        }
    }

    */
/**
     * use [.setCallbackIfNotSet] instead
     * @param callback
     *//*

    @Deprecated("")
    fun setCallback(callback: DiagnosticsCallback) {
        setCallbackIfNotSet(callback)
    }

    fun setCallbackIfNotSet(callback: DiagnosticsCallback): Boolean {
        return false
    }

    fun resetCallback() {}

    fun wantsDiagnostics(): Boolean
}


*/
